urlaa051fbd3d_WheezePuppet_collegeDiversity_tgz-pJ8-edu_umw_cpsc_collegediversity_SimJ8/src/edu/umw/cpsc/collegediversity/UpperclassHousingSelection.java

```diff
@@ -0,0 +1,246 @@
+
+        //Willard houses students in single rooms which is not accounted for yet
+        Dorm Willard = new Dorm("Willard", false, false, 91); 
+
+        //100 of the rooms in the aparments are singles which is not 
+        // accounted for yet
+        Dorm Apartments = new Dorm("Apartments", false, false, 381); 
+
+        //for extraneous students who commute or live off campus
+        Dorm Offcampus = new Dorm ("Offcampus", false, false, 3500); 
+
+        dorms.add(Ball);
+        dorms.add(Arrington);
+        dorms.add(Custis);
+        dorms.add(Framar);
+        dorms.add(Madison);
+        dorms.add(Mason);
+        dorms.add(Westmoreland);
+        dorms.add(Willard);
+        dorms.add(Apartments);
+        dorms.add(EagleLanding);
+        dorms.add(Offcampus);
+    }
+
+    /**
+     * Singleton pattern.
+     */
+    public static UpperclassHousingSelection instance(){
+        if(theInstance==null){
+            theInstance=new UpperclassHousingSelection();
+        }
+        return theInstance;
+    }
+
+
+    /** 
+     * Assign the set of students (presumably non-incoming-freshmen, but
+     * this is not checked) passed to their dorm rooms, in a way that does 
+     * not take race into account. (compare {@link #assignByRace}.) As a 
+     * precursor (side effect) to this, any students already existing in 
+     * upperclass dorms will be removed.
+     */
+    public void assign(Bag students){
+
+        emptyDorms();
+
+        // Purge all freshmen. (We're not assigning those.)
+        Bag upperclassmen = null;
+        try { 
+            upperclassmen = (Bag) students.clone(); 
+        } catch (Exception e) { 
+            e.printStackTrace(); 
+            System.exit(1); 
+        }
+        for(int x=upperclassmen.size()-1; x>=0; x--){
+            Student s = (Student) upperclassmen.get(x);
+            if (s.getGrade() < 2) {
+                upperclassmen.remove(s);
+            }
+        }
+        
+        for(int x=0;x<upperclassmen.size();x++){
+            Student s = (Student) upperclassmen.get(x);
+            if (s.getGrade() < 2) {
+                // We're only assigning upperclassmen here.
+                continue;
+            }
+            s.leaveRoom();
+            for(int y=0; y<dorms.size();y++){
+                if(s.hasRoom()){break;}
+                Dorm d = (Dorm) dorms.get(y);
+                if(d.isFull()){
+                    continue;
+                }
+                if(d.isFemaleOnly() && s.getGender()==Student.Gender.MALE){
+                    continue;
+                }
+                int i = 0;
+                while(i<d.getNumRooms()){
+                    if(s.hasRoom()){break;}
+                    try{
+                        d.getRoomByIndex(i).addResident(s);
+                    }catch(IllegalArgumentException e){//System.out.println(e);
+                    }
+                    i++;
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 103:121, 217:236.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.